crates/runtime/src/memory.rs

```diff
@@ -254,35 +260,136 @@ impl RuntimeLinearMemory for MmapMemory {
         // is needed.
         self.memory_image.is_none()
     }
+
+    #[cfg(feature = "pooling-allocator")]
+    fn into_any(self: Box<Self>) -> Box<dyn std::any::Any> {
+        self
+    }
 }
 
-/// Representation of a runtime wasm linear memory.
-pub enum Memory {
-    /// A "static" memory where the lifetime of the backing memory is managed
-    /// elsewhere. Currently used with the pooling allocator.
-    Static {
-        /// The memory in the host for this wasm memory. The length of this
-        /// slice is the maximum size of the memory that can be grown to.
-        base: &'static mut [u8],
+/// A "static" memory where the lifetime of the backing memory is managed
+/// elsewhere. Currently used with the pooling allocator.
+pub(crate) struct ExternalMemory {
+    /// The memory in the host for this wasm memory. The length of this
+    /// slice is the maximum size of the memory that can be grown to.
+    base: &'static mut [u8],
 
-        /// The current size, in bytes, of this memory.
-        size: usize,
+    /// The current size, in bytes, of this memory.
+    size: usize,
 
-        /// A callback which makes portions of `base` accessible for when memory
-        /// is grown. Otherwise it's expected that accesses to `base` will
-        /// fault.
-        make_accessible: Option<fn(*mut u8, usize) -> Result<()>>,
+    /// A callback which makes portions of `base` accessible for when memory
+    /// is grown. Otherwise it's expected that accesses to `base` will
+    /// fault.
+    make_accessible: Option<fn(*mut u8, usize) -> Result<()>>,
+
+    /// The image management, if any, for this memory. Owned here and
+    /// returned to the pooling allocator when termination occurs.
+    pub(crate) memory_image: Option<MemoryImageSlot>,
+}
 
-        /// The image management, if any, for this memory. Owned here and
-        /// returned to the pooling allocator when termination occurs.
+impl ExternalMemory {
+    fn new(
+        base: &'static mut [u8],
+        size: usize,
+        make_accessible: Option<fn(*mut u8, usize) -> Result<()>>,
         memory_image: Option<MemoryImageSlot>,
-    },
+    ) -> Result<Self> {
+        if base.len() < size {
+            bail!(
+                "initial memory size of {} exceeds the pooling allocator's \
+                 configured maximum memory size of {} bytes",
+                size,
+                base.len(),
+            );
+        }
 
-    /// A "dynamic" memory whose data is managed at runtime and lifetime is tied
-    /// to this instance.
-    Dynamic(Box<dyn RuntimeLinearMemory>),
+        if let Some(make_accessible) = make_accessible {
+            if size > 0 {
+                make_accessible(base.as_mut_ptr(), size)?;
+            }
+        }
+
+        Ok(Self {
+            base,
+            size,
+            make_accessible,
+            memory_image,
+        })
+    }
+}
+
+impl Default for ExternalMemory {
```

Comment:
I think this may actually be historical at this point, while you're here could you try removing this (and dependants) and see if it's actually still needed?

Reply:
I tried... `mem::take` in `pooling.rs` requires a `Memory` to be `Default`. Is there another way to do something equivalent there?

Reply:
Ah that can be replaced with `mem::take` of the `memories` list itself rather than each individual memory I thin

Reply:
Can we? `memory` is coming from `self.memories.get(instance_index)` so I don't think I can just take from the iterator? Maybe I'm just not seeing it...

Reply:
Yes with `memory: &mut PrimaryMap<K, V>` you can `mem::take`. Only the `base` comes from `self.memories.get(instance_index)`

Reply:
Oh, you're right! Sorry, wasn't looking closely enough.